src/core/net/srp_client.hpp

```diff
@@ -0,0 +1,731 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SRP_CLIENT_HPP_
+#define SRP_CLIENT_HPP_
+
+#include "openthread-core-config.h"
+
+#include <openthread/srp_client.h>
+
+#include "common/clearable.hpp"
+#include "common/linked_list.hpp"
+#include "common/locator.hpp"
+#include "common/message.hpp"
+#include "common/non_copyable.hpp"
+#include "common/notifier.hpp"
+#include "common/timer.hpp"
+#include "crypto/ecdsa.hpp"
+#include "net/dns_headers.hpp"
+#include "net/ip6.hpp"
+#include "net/udp6.hpp"
+
+#if OPENTHREAD_CONFIG_SRP_CLIENT_ENABLE
+
+/**
+ * @file
+ *   This file includes definitions for the SRP (Service Registration Protocol) client.
+ */
+
+namespace ot {
+namespace Srp {
+
+#if !OPENTHREAD_CONFIG_ECDSA_ENABLE
+#error "SRP Client feature requires ECDSA support (OPENTHREAD_CONFIG_ECDSA_ENABLE)."
+#endif
```

Comment:
I prefer/suggest such checks to be included in the header file of the module itself. I think it's better to keep the checks close to the code itself.  
-  It will help reader of the code (since the config dependency checks are close to code/definition itself)
- Being in the same header file, can help developers not forget updating the list if/when new code is added which adds new dependencies.
- The `config-check.h` header was initially intended for removed/renamed or deprecated config options.

Reply:
It seems `openthread-core-config-check.h` now goes beyond its original purpose, and also serves as a central place for the dependencies checks. 

I think it would be better to follow existing style, and may use a separate PR to restore the original intention of the `openthread-core-config-check.h` if you prefer to keep the dependencies into separate components?

Reply:
I prefer to keep these here. 

I think the fact that we started using `config-check.h` for these dependency check was not good decision in first place. There are so many checks in `config-check` header (all related to renamed/deprecated configs) which makes it very hard to read and follow what is being checked. It does not make sense to me to now add more checks in it.

In terms of precedence, we already have cases where dependencies are checked in other files.
- `core/common/logging.cpp`:#error OPENTHREAD_CONFIG_ENABLE_DEBUG_UART_LOG requires OPENTHREAD_CONFIG_ENABLE_DEBUG_UART
- `core/common/message.cpp`:#error "OPENTHREAD_CONFIG_MESSAGE_USE_HEAP_ENABLE is strongly discouraged when OPENTHREAD_CONFIG_DTLS_ENABLE is off."
- `core/thread/network_data_leader.hpp`:#error "Please define OPENTHREAD_MTD=1 or OPENTHREAD_FTD=1"
- `core/thread/topology.hpp`:#error OPENTHREAD_CONFIG_MLE_IP_ADDRS_PER_CHILD should be at least set to 2.
- `core/config/backbone_router.h`:#error 
- `core/config/tmf.h`:#error "Thread 1.2 or higher version is required for OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE"
- `core/config/backbone_router.h` - multiple cases here.